### PR TITLE
PyStrKind to optimize ascii strings

### PR DESCRIPTION
**Pull request by youknowone**
And this structure also can be used for invalid unicode strings and ucs16/ucs32 representations in CPython in future.

### CODE REVIEW

File: vm/src/builtins/pystr.rs [hunk 1]
```diff
@@ -294,44 +339,52 @@ impl PyStr {
     }
 
     pub fn as_str(&self) -> &str {
-        &self.value
```
Comment (DimitrisJim) [inline vm/src/builtins/pystr.rs:297]: Safety comment? I'm haven't seen a specific rule throughout the project but I'm guessing comments describing safety should be required.

Also, wouldn't this benefit from being marked as `#[inlined(always)]?

Comment (youknowone) [reply vm/src/builtins/pystr.rs:297]: I added safey comments for every unsafe. About inline, I added one but I also tihnk these kind of tiny functions are easily marked as inline by compiler

Comment (DimitrisJim) [reply vm/src/builtins/pystr.rs:297]: [Reading a bit about this](https://internals.rust-lang.org/t/inlining-policy-for-functions-in-std/14189/10), for private functions you're probably right. Cross crate public ones don't automatically get inlined. It's probably not harmful to add in this case.

**Review by DimitrisJim — APPROVED** [review]

**Comment by coolreader18** [comment]
This is unsound - you can pass a non-ascii string to new_ascii_str, and a non-utf8 string to `::from((v, Kind::Utf8))` that'll get passed without validation to `unsafe { str::from_utf8_unchecked }`

**Pull request merged**